File: C++/A1146 Topological Order.cpp
#include <iostream>
#include <string>
#include <vector>
#include <iomanip>
#include <algorithm>

//前两个和最后一个测试点错误
int main() {
	int n;
	int m;
	scanf_s("%d",&n);
	scanf_s("%d",&m);
	
	bool map[1000][1000] = { false };
	for (int i = 0; i < m; i++) {
		int from;
		int to;
		scanf_s("%d", &from);
		scanf_s("%d", &to);
		map[from-1][to-1] = true;
	}
	int k;
	scanf_s("%d", &k);
	bool notFirst = false;
	for (int i = 0; i < k; i++) {
		bool visit[1000] = { false };
		for (int j = 0; j < n; j++) {
			int now;
			scanf_s("%d", &now);
			visit[now-1] = true;
			for (int k1 = 0; k1 < n; k1++) {
				if (map[k1][now-1] && !visit[k1]) {
					if (notFirst) printf(" ");
					printf("%d",i);
					notFirst = true;
					goto loop;
				}
			}
		}
	loop:int x = 0;
	}

	getchar();
	getchar();

	return 0;
}

File: java/A1146 Topological Order.java
//前两个测试点答案错误，最后一个测试点超时
import java.util.Scanner;

public class Main {

	public static void main(String[] args) {
		Scanner scanner = new Scanner(System.in);
		int n = scanner.nextInt();
		int m = scanner.nextInt();

		boolean[][] map = new boolean[n][n];
		
		for(int i=0;i<m;i++) {
			int from = scanner.nextInt()-1;
			int to = scanner.nextInt()-1;
			map[from][to]=true;
		}
		
		int k = scanner.nextInt();
		boolean notFirst = false;
		loop:
		for(int i=0;i<k;i++) {
			boolean[] visit = new boolean[n];
			for(int j=0;j<n;j++) {
				int now = scanner.nextInt()-1;
				visit[now]=true;
				for(int k1=0;k1<n;k1++) {
					if(map[k1][now]&&!visit[k1]) {
						if(notFirst) System.out.print(" ");
						System.out.print(i);
						notFirst = true;
						continue loop;
					}
				}
			}
		}
		
		scanner.close();
	}
}